Arm Actuator Interface

niiquaye opened the pull request:
First Commit for Arm Actuator Interface

meshvaD committed:
urdf arm description

wmmc88 committed:
Windows CI + misc ci improvements (#156)

* update dependabot reviewers
* fix false positive on arm64 test
* fix image selection
* add global package list
* use ros galactic
* workaround for upstream windows bug (Imported target "rclcpp::rclcpp" includes non-existent path)
* use variable in image field

meshvaD committed:
launch

niiquaye committed:
Arm Actuator Interface

niiquaye commented:
Going to work on finishing up the system interface that will represent a5/a6 with the differential transmission - that will be my next commit

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+
+add_custom_target(
+        copy-compile-commands ALL
+        ${CMAKE_COMMAND} -E copy_if_different
+        ${CMAKE_BINARY_DIR}/compile_commands.json
+        ${CMAKE_CURRENT_LIST_DIR}
+)
+
+
+
```
What is this for?

niiquaye replied:
oh this was when I was trying to get my compile_commands.json to get produced ..... long story short I got it work and the cause of the issue was something else entirely .... I forgot to remove this

wmmc88 replied:
the .clang-format file is already in the main branch of the repo. see how i setup this cmake function to use it in #171

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+find_package(rclcpp REQUIRED)
+find_package(rclcpp_lifecycle REQUIRED)
+
+include_directories(include)
```
Should use target based includes instead

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+
+  # clang-format
+  find_package(ament_cmake_clang_format REQUIRED)
+  ament_clang_format()
```
Needs our .clangformat file

niiquaye replied:
yes we def need this ..... it'll make our formatting standard when we run .... ament_clang_format <folders> --reformat ..... we can just copy and paste the old one from the ros1 repo right

wmmc88 replied:
The main branch of the repo already has the .clang-format file. see #171 to see how to call the cmake function to specify the config

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+  ament_clang_tidy(${CMAKE_BINARY_DIR})
+endif()
+
+ament_package()
```
Rewrite this file to match drivetrain hw I'm 171 imo

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/uwrt_mars_rover_arm_hw_actuator.hpp:
```diff
@@ -0,0 +1,147 @@
+
+namespace uwrt_mars_rover_arm_hw
+{
+// namespaced structs to hold transmission data - state & command
```
Redundant comment. States obvious

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/uwrt_mars_rover_arm_hw_actuator.hpp:
```diff
@@ -0,0 +1,147 @@
+  double velocity{};
+  double position{};
+  double
+    iq_current{};  // actuator also provide iq_current - this is not needed for transmission however
```
```suggestion
    iq_current{};  // unused in transmission
```

More succinct comments

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/uwrt_mars_rover_arm_hw_actuator.hpp:
```diff
@@ -0,0 +1,147 @@
+    const rclcpp_lifecycle::State & previous_state) override;  // configuration
+
+  UWRT_MARS_ROVER_ARM_HW_PUBLIC
+  CallbackReturn on_activate(const rclcpp_lifecycle::State & previous_state) override;  // start
```
Remove the comment on these lifecycle functions

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/uwrt_mars_rover_arm_hw_actuator.hpp:
```diff
@@ -0,0 +1,147 @@
+  inline static constexpr std::size_t NUM_COMMAND_INTERFACES{1};
+
+  // holds actuator reads/writes
+  TransmissionData::ActuatorData::CommandData actuator_commands{};
+  TransmissionData::ActuatorData::StateData actuator_states{};
+  TransmissionData::JointData::CommandData joint_commands{};
+  TransmissionData::JointData::StateData joint_states{};
```
Remove transmission stuff from this pr and address seperately

niiquaye replied:
can I keep it pretty pretty pretty please .... I'm not using any transmission stuff "really" .... its really just a place holder of how it would be set up ..... I only use the data members of my defined structs to add to the command & state interfaces

wmmc88 replied:
if its truly unused, then clang-tidy will prevent you from merging it anyways. I would just stash these lines for a future pr

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/uwrt_mars_rover_arm_hw_actuator.hpp:
```diff
@@ -0,0 +1,147 @@
+
+  //std::shared_ptr<rclcpp::Logger> actuator_logger {nullptr};
+
+  // iq_current interface
+  static constexpr char HW_IF_IQ_CURRENT[] = "iq_current";
```
See discussion in #171

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/uwrt_mars_rover_arm_hw_actuator.hpp:
```diff
@@ -0,0 +1,147 @@
+
+  double joint_to_actuator_reduction{};
+
+  // create a logger that will be unique with any joint name
+  std::function<rclcpp::Logger(void)> actuator_logger{};  //use logger
```
See discussion in #171

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/visibility.hpp:
```diff
@@ -0,0 +1,65 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
```
Remove this since it's not copying their header, just basing off the gcc wiki entry. 

Also fyi our repo is MIT licensed

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/visibility.hpp:
```diff
@@ -0,0 +1,65 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef UWRT_MARS_ROVER_ARM_HW__VISIBILITY_H_
```
Pragma once

niiquaye replied:
damn idk how I missed that

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/visibility.hpp:
```diff
@@ -0,0 +1,65 @@
+#ifndef UWRT_MARS_ROVER_ARM_HW__VISIBILITY_H_
+#define UWRT_MARS_ROVER_ARM_HW__VISIBILITY_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Remove. We have no intent to use c

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/visibility.hpp:
```diff
@@ -0,0 +1,65 @@
+#define UWRT_MARS_ROVER_ARM_HW_IMPORT __declspec(dllimport)
+#endif
+
+#ifdef UWRT_MARS_ROVER_ARM_HW_DLL
```
This should be BUILDING_LIBRARY to match ros conventions. See #171

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/include/uwrt_mars_rover_arm_hw/visibility.hpp:
```diff
@@ -0,0 +1,65 @@
+}
+#endif
+
+#endif  // UWRT_MARS_ROVER_ARM_HW__VISIBILITY_H_
```
This file should be visibility_control.h to match ros conventions

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/package.xml:
```diff
@@ -0,0 +1,29 @@
+<package format="3">
+  <name>uwrt_mars_rover_arm_hw</name>
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
```
Fix todos

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/package.xml:
```diff
@@ -0,0 +1,29 @@
+  <version>0.0.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="omarmon@uwaterloo.ca">niiquaye</maintainer>
+  <license>TODO: License declaration</license>
```
Fix todos

wmmc88 commented on uwrt_mars_rover_arm/uwrt_mars_rover_arm_hw/src/uwrt_mars_rover_arm_hw_actuator.cpp:
```diff
@@ -0,0 +1,376 @@
+
+#include <cstdlib>
+#include <hardware_interface/actuator_interface.hpp>
+#include <hardware_interface/handle.hpp>
+#include <hardware_interface/hardware_info.hpp>
+#include <hardware_interface/types/hardware_interface_return_values.hpp>
+#include <hardware_interface/types/hardware_interface_type_values.hpp>
+#include <limits>
+#include <rclcpp/logger.hpp>
+#include <rclcpp/logging.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp_lifecycle/node_interfaces/lifecycle_node_interface.hpp>
```
I prefer all ros includes to be quote included. Afaik, that's ros convention

wmmc88 commented on uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_hw/launch/drivetrain.launch.py:
```diff
@@ -0,0 +1,54 @@
+
+    robo = (
+        Path(
+            r"/home/niiquaye/robotics/install/uwrt_mars_rover_drivetrain_description/share/uwrt_mars_rover_drivetrain_description/urdf/robot.urdf"
```
This is not portable. See launch files in #171

niiquaye replied:
oh ya I gotta remove this .... idk why I even have this in this commit

wmmc88 commented:
Add relevant packages to the workflow

meshvaD committed:
checks

niiquaye commented on .gitignore:
```diff
@@ -12,6 +12,10 @@ msg/*Result.msg
 msg/_*.py
 build_isolated/
 devel_isolated/
+compile_commands.json
+.clangd
+
+.ccls-cache/
```
from me .... some ide config stuff

wmmc88 replied:
compile_commands shouldnt be generated in a folder thats not already .git-ignored, so if it is, its an indication that you invoked the command from the wrong directory or set something else up wrong

niiquaye replied:
ya I was going pretty ham on trying to generate it ..... it turned out I did not properly setup my executable so the compile_commands were not coming out right ..... but that was a completely different issue

niiquaye committed:
System Interface Header File

niiquaye committed:
Arm Differential Transmission Implementation File
*defined functions in the extended system interface class
*use Logger object to log info

niiquaye commented:
Accidentally committed some unwanted files .... also gotta fix some merge conflicts

niiquaye committed:
Fixing Stuff

niiquaye committed:
Fixing Stuff pt2
